### ISSUE TITLE
matches: an invalid regex is silently false, and patterns are recompiled on every call

### ISSUE THREAD

**Issue by PaulSnow** [DTRules/DTRules#1262]
Found in review of #1235 (PR for #1228). This is the same "silently false" class as #1228.

## An invalid pattern is silently false

`opRegexMatch` (`pkg/dtrules/operators/string.go`, ~line 316):

```go
matched, err := regexp.MatchString(pattern, str)
if err != nil {
    return state.DataPush(dtrules.GetRBoolean(false))
}
```

A pattern that does not compile, such as `s matches "("` or `s matches "[a-"`, evaluates to **false**. Nothing is raised, logged or warned. A mistyped regex therefore reads as "does not match": the rule takes its N branch and gives a plausible wrong answer. Every other operator in this family returns an error for bad input.

## No compile cache

`regexp.MatchString` compiles the pattern on every call. A `matches` in a condition that is evaluated per element of a large array (inside `for all …`) recompiles the same literal thousands of times.

## What `matches` means is undocumented

Go's `MatchString` is an **unanchored search**: `"xabcx" matches "abc"` is true. The EL reference says only "True if the string matches the regular expression". An author coming from Java DTRules (`String.matches`, which is a full match) or from XSD patterns would expect `"xabcx" matches "abc"` to be false.

## Proposal

1. An invalid pattern returns an error from `regexmatch`, naming the pattern and the regexp error, so the rule fails where the mistake is.
2. Compiled patterns are cached per pattern string (a `sync.Map` of `*regexp.Regexp`). Patterns in rules are overwhelmingly literals, so the cache stays small.
3. The docs (`docs/el-reference.md`, `dtrules docs el`) say that `matches` is an unanchored search, and that `^…$` gives a full match.

   Alternative: decide on full-match semantics. That would silently change the result of every existing unanchored pattern, the same failure mode as #1228, so I recommend documenting the current behaviour instead.
